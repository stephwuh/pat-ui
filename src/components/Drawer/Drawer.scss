.drawer{
    background-color: rgb(18, 18, 18);
    color: rgb(255, 255, 255);
    transition: box-shadow 300ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
    box-shadow: rgb(0 0 0 / 20%) 0px 8px 10px -5px, rgb(0 0 0 / 14%) 0px 16px 24px 2px, rgb(0 0 0 / 12%) 0px 6px 30px 5px;
    background-image: linear-gradient(rgba(255, 255, 255, 0.15), rgba(255, 255, 255, 0.15));
    height: 100%;
    position: fixed;
    padding: 30px;
}

.drawer-left{
    top: 0;
    left: 0;
    height: 100%;
    width: max-content;
}

.drawer-right{
    top: 0;
    right: 0;
    height: 100%;
    width: max-content;
}

.drawer-top{
    top:0;
    width: 100%;
    height: max-content;
}

.drawer-bottom{
    bottom:0;
    width: 100%;
    height: max-content;
}

.drawer-openStateFalse{
    display: none;
}

.drawer-dimBackground{ 
    position: fixed;
    display: flex;
    -webkit-box-align: center;
    align-items: center;
    -webkit-box-pack: center;
    justify-content: center;
    inset: 0px;
    background-color: rgba(0, 0, 0, 0.5);
    -webkit-tap-highlight-color: transparent;
    z-index: -1;
}