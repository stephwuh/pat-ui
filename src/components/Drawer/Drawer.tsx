import React, { FC, useState } from 'react';
import { classNames } from '../../utils/classNames';

type Variant = 'permanent' | 'persistent' | 'temporary';

type Anchor = 'top' | 'bottom' | 'left' | 'right';

export interface IDrawerProps {
    open?: boolean;
    anchor?: Anchor;
    className?: string;
    variant?: Variant;
    onToggleCallback?: () => void;
}

/**
 * A drawer allows user to hide or show a sidebar with annimations.
 *
 * ```js
 * import { Drawer } from 'pat-ui'
 * ```
 */

const Drawer: FC<IDrawerProps> = ({
    open = false,
    anchor = 'left',
    variant = 'temporary',
    children,
    className,
    onToggleCallback
}) => {

    const handleToggleDrawer = (event: React.MouseEvent) => {
        if (onToggleCallback) {
            onToggleCallback();
        }
    };

    let styleClasses = classNames('drawer', {
        [`${variant === 'permanent' ? `drawer-${anchor}` : ''}`]: true,
        [`${variant !== 'permanent' && open ? `drawer-open-${anchor}` : ''}`]: true,
        [`${open ? '' : `drawer-close-${anchor}`}`]: true,
        [`${variant === 'permanent' ? `drawer-permanent` : ''}`]: true,
        [`${variant === 'persistent' ? `drawer-persistent` : ''}`]: true,
        [`${variant === 'temporary' ? `drawer-temporary` : ''}`]: true,
    })
    if (className) {
        styleClasses += ' ' + className;
    }

    let dimBackgroundStyle = '';
    if (variant === 'temporary' && open === true) {
        dimBackgroundStyle = 'drawer-dimBackground';
    }

    const [closeDrawerStyle, setCloseDrawerStyle] = useState<string>('');

    setTimeout(() => {
        let closeDrawer = `${open ? 'drawer-open' : 'drawer-close'}`;
        setCloseDrawerStyle(closeDrawer);
    }, 300)

    return (
        <section className={closeDrawerStyle}>
            <div className={dimBackgroundStyle} onClick={handleToggleDrawer} data-testid='dimmed-background'></div>
            <aside className={styleClasses} data-testid="drawer">
                {children}
            </aside>
        </section>
    )
}

export default Drawer;