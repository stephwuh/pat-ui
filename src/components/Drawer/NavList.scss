.list-item__icon{
    padding: 15px;
    position: relative;
    top: 5px;
}

.navList{
    font-size: 1.2em;
}

.list-item__label{
    padding: 0 15px 0 0;
}