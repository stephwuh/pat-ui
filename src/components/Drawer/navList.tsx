import React, { FC, useState, useEffect } from 'react';


interface INavListProps {
    listItemLables?: string[];
    listItemIcons?: any[];
    listItemLinkPaths?: string[];

    listItemData?: { name: string, icon: any, linkPath: string };
}

const NavList: FC<INavListProps> = ({
    listItemLables,
    listItemIcons,
    listItemLinkPaths,

}) => {

    const [listItemData, setListItemData] = useState([{name:'',icon:'',linkPath:''}]);
    const newData = [{name:'', icon:'', linkPath:''}];

    useEffect(() => {
        let i: any;
        for (i in listItemLables) {
            const data = { name: listItemLables![i], icon: listItemIcons![i], linkPath: listItemLinkPaths![i] };
            newData.push(data);
        }   
        newData.shift();
        setListItemData(newData);
    }, []);

    return (
        <section className='navList'>
            {listItemData?.map((listItem, i) => [
                <h1 key={i} className='list-item__label'>
                <span className='list-item__icon'>{listItem.icon}</span>
                {listItem.name}</h1>
            ])}
        </section>
    )
}

export default NavList;