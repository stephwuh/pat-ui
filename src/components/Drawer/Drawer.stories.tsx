import React, { useState } from 'react';
import Drawer from './Drawer';

export default {
    title: 'Drawer',
    component: Drawer
}

export const permanentDrawer = () => (
    <div>
        <Drawer
            open={true}
            variant={'permanent'}
        >
            <h2>Inbox</h2>
            <h2>All Mail </h2>
            <h2>Send Email</h2>
            <hr></hr>
            <h2>Spam</h2>
            <h2>Trash</h2>
        </Drawer>
    </div>
)

export const persistentDrawer = () => {

    const [isOpen, setIsOpen] = useState<boolean>(false);

    const handleToggle = () => {
        setIsOpen(!isOpen);
    }

    return (
        <div style={{ display: 'flex', columnGap: '50px' }}>
            <Drawer
                open={isOpen}
                variant={'persistent'}
                onToggleCallback={handleToggle}
            >
                <h2>Inbox</h2>
                <h2>All Mail </h2>
                <h2>Send Email</h2>
                <hr></hr>
                <h2>Spam</h2>
                <h2>Trash</h2>
            </Drawer>
            <button style={{ height: '50px', alignSelf: 'center' }} onClick={handleToggle}>Open / Close </button>
        </div>
    );
};

export const temporaryDrawer = () => {

    const [isOpen, setIsOpen] = useState<boolean>(false);

    const handleToggle = () => {
        setIsOpen(!isOpen);
    }

    return (
        <div style={{ display: 'inline-block' }}>
            <Drawer
                open={isOpen}
                variant={'temporary'}
                onToggleCallback={handleToggle}
            >
                <h2>Inbox</h2>
                <h2>All Mail </h2>
                <h2>Send Email</h2>
                <hr></hr>
                <h2>Spam</h2>
                <h2>Trash</h2>
            </Drawer>
            <button style={{ height: '40px', alignSelf: 'center', margin: '0 50px' }} onClick={handleToggle}> Open </button>
            <i> Click dimmed background to close </i>
        </div>
    )
}

export const LeftDrawer = () => (
    <div>
        <Drawer
            open={true}
            anchor={'left'}
            variant={'persistent'}
        >
            <h2>Inbox</h2>
            <h2>All Mail </h2>
            <h2>Send Email</h2>
            <hr></hr>
            <h2>Spam</h2>
            <h2>Trash</h2>
        </Drawer>
    </div>
)

export const RightDrawer = () => (
    <div>
        <Drawer
            open={true}
            anchor={'right'}
            variant={'persistent'}
        >
            <h2>Inbox</h2>
            <h2>All Mail </h2>
            <h2>Send Email</h2>
            <hr></hr>
            <h2>Spam</h2>
            <h2>Trash</h2>
        </Drawer>
    </div>
)

export const TopDrawer = () => (
    <div>
        <Drawer
            open={true}
            anchor={'top'}
            variant={'persistent'}
        >
            <h2>Inbox</h2>
            <h2>All Mail </h2>
            <h2>Send Email</h2>
            <hr></hr>
            <h2>Spam</h2>
            <h2>Trash</h2>
        </Drawer>
    </div>
)

export const BottomDrawer = () => (
    <div>
        <Drawer
            open={true}
            anchor={'bottom'}
            variant={'persistent'}
        >
            <h2>Inbox</h2>
            <h2>All Mail </h2>
            <h2>Send Email</h2>
            <hr></hr>
            <h2>Spam</h2>
            <h2>Trash</h2>
        </Drawer>
    </div>
)