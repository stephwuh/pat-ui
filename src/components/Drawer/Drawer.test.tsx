import React from 'react';

import Drawer from './Drawer';

import { render, screen } from '@testing-library/react';

describe('Drawer', () => {
  it('should match snapshot', () => {
    const { asFragment } = render(<Drawer />);
    expect(asFragment()).toMatchSnapshot();
  });
  it('Developers can choose the style of the temporary Drawer using props', () => {
    render(<Drawer variant="temporary" />);
    const drawerElement = screen.getByTestId('drawer');
    expect(drawerElement).toHaveClass('drawer-temporary');
    expect(drawerElement).not.toHaveClass('drawer-persistent');
  });
  it('Developers can choose the style of the persistent Drawer using props', () => {
    render(<Drawer variant="persistent" />);
    const drawerElement = screen.getByTestId('drawer');
    expect(drawerElement).toHaveClass('drawer-persistent');
    expect(drawerElement).not.toHaveClass('drawer-temporary');
  });

  it('Developers can control the open state of the component from props.', () => {
    render(<Drawer open={true} />);
    const drawerElement = screen.getByTestId('drawer');
    expect(drawerElement).toHaveClass('drawer-open-left');
    expect(drawerElement).not.toHaveClass('drawer-close-left');
  });

  it('Developers can control the close state of the component from props.', () => {
    render(<Drawer open={false} />);
    const drawerElement = screen.getByTestId('drawer');
    expect(drawerElement).toHaveClass('drawer-close-left');
    expect(drawerElement).not.toHaveClass('drawer-open-left');
  });

  it('Developers should see a dimed backdrop when the drawer component is open when style is set to temporary.', () => {
    render(<Drawer open={true} variant="temporary" />);
    const drawerElement = screen.getByTestId('dimmed-background');
    expect(drawerElement).toHaveClass('drawer-dimBackground');
  });

  it('Developers should not see a dimed backdrop when the drawer component is closed.', () => {
    render(<Drawer open={false} variant="temporary" />);
    const drawerElement = screen.getByTestId('dimmed-background');
    expect(drawerElement).not.toHaveClass('drawer-dimBackground');
  });

  it('Developers should not see a dimed backdrop when the drawer component style is set to persistent.', () => {
    render(<Drawer open={true} variant="persistent" />);
    const dimmedBackgroundDivElement = screen.getByTestId('dimmed-background');
    expect(dimmedBackgroundDivElement).not.toHaveClass('drawer-dimBackground');
  });

  it('Developers should be able to choose the left open direction of the drawer component from props.', () => {
    render(<Drawer open={true} anchor="left" />);
    const drawerElement = screen.getByTestId('drawer');
    expect(drawerElement).toHaveClass('drawer-open-left');
    expect(drawerElement).not.toHaveClass('drawer-open-right');
    expect(drawerElement).not.toHaveClass('drawer-open-top');
    expect(drawerElement).not.toHaveClass('drawer-open-bottom');
  });

  it('Developers should be able to choose the right open direction of the drawer component from props.', () => {
    render(<Drawer open={true} anchor="right" />);
    const drawerElement = screen.getByTestId('drawer');
    expect(drawerElement).toHaveClass('drawer-open-right');
    expect(drawerElement).not.toHaveClass('drawer-open-left');
    expect(drawerElement).not.toHaveClass('drawer-open-top');
    expect(drawerElement).not.toHaveClass('drawer-open-bottom');
  });

  it('Developers should be able to choose the top open direction of the drawer component from props.', () => {
    render(<Drawer open={true} anchor="top" />);
    const drawerElement = screen.getByTestId('drawer');
    expect(drawerElement).toHaveClass('drawer-open-top');
    expect(drawerElement).not.toHaveClass('drawer-open-left');
    expect(drawerElement).not.toHaveClass('drawer-open-right');
    expect(drawerElement).not.toHaveClass('drawer-open-bottom');
  });

  it('Developers should be able to choose the bottom open direction of the drawer component from props.', () => {
    render(<Drawer open={true} anchor="bottom" />);
    const drawerElement = screen.getByTestId('drawer');
    expect(drawerElement).toHaveClass('drawer-open-bottom');
    expect(drawerElement).not.toHaveClass('drawer-open-left');
    expect(drawerElement).not.toHaveClass('drawer-open-right');
    expect(drawerElement).not.toHaveClass('drawer-open-top');
  });

  it('Developers can listen to the change of the close event of the component by providing the onClose callback to the props.', () => {
    let openState = true;
    const handleClick = jest.fn().mockImplementation(() => {
      openState = !openState;
    });

    render(
      <div>
        <button onClick={handleClick()}>Close</button>
        <Drawer open={openState} onToggleCallback={handleClick} />
      </div>
    );
    const drawerElement = screen.getByTestId('drawer');
    expect(drawerElement).toHaveClass('drawer-close-left');
  });


});