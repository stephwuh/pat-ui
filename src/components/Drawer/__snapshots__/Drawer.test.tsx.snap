// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Drawer should match snapshot 1`] = `
<DocumentFragment>
  <section
    class=""
  >
    <div
      class=""
      data-testid="dimmed-background"
    />
    <aside
      class="drawer  drawer-close-left drawer-temporary"
      data-testid="drawer"
    />
  </section>
</DocumentFragment>
`;
